### PR TITLE
Move span operation to Scope from Tracer

### PR DESCRIPTION
**Pull request by suranjay**
### Description
This PR makes following changes to Tracing framework
- Move span operation to Scope from Tracer
- Add span status

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by suranjay** [comment]
@reta Please have a look of the PR when you get chance

File: plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelSpan.java [hunk 1]
```diff
@@ -48,6 +48,11 @@ public void addAttribute(String key, Boolean value) {
         delegateSpan.setAttribute(key, value);
     }
 
+    @Override
+    public void setStatus(StatusCode statusCode) {
+        delegateSpan.setStatus(io.opentelemetry.api.trace.StatusCode.valueOf(statusCode.name()));
```
Comment (suranjay) [inline plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelSpan.java:53]: I think we should have mapper(OS status code->OTel status code) here

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java [hunk 2]
```diff
@@ -62,6 +62,13 @@ public interface Span {
      */
     void addAttribute(String key, Boolean value);
 
+    /**
+     * Sets status of the span
+     *
+     * @param statusCode status of the span
+     */
+    void setStatus(StatusCode statusCode);
```
Comment (reta) [inline libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java:70]: I don't think we need the setStatus and StatusCode: the typical way the errors are reported in Java are through exceptions, so we need `error` / `setError` to propagate that.

```
try (Scope s = tracer.startSpan(...)) {
  ...
} catch (Exception ex) {
  s.error(ex);
}
```

**Comment by reta** [comment]
@suranjay @Gaganjuneja seems like folks you are have conflicting changes https://github.com/opensearch-project/OpenSearch/pull/8357

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java (hunk 2 above)
Comment (suranjay) [reply libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java:70]: Sounds reasonable as we have mainly two statuses error and non error. 
So in our Scope implementation we save it as status in Otel ?
```
public void setStatus(Exception e) {
        delegateSpan.setStatus(StatusCode.ERROR, e.getMessage());
    }
```

**Comment by suranjay** [comment]
> @suranjay @Gaganjuneja seems like folks you are have conflicting changes #8357

Yes, we will rebase the other PR changes which gets merged earlier

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java (hunk 2 above)
Comment (suranjay) [reply libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/Span.java]: Addressed in latest commit

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java [hunk 3]
```diff
@@ -36,41 +36,11 @@ public DefaultTracer(TracingTelemetry tracingTelemetry, TracerContextStorage<Str
     }
 
     @Override
-    public Scope startSpan(String spanName) {
+    public SpanScope startSpan(String spanName) {
         Span span = createSpan(spanName, getCurrentSpan());
         setCurrentSpanInContext(span);
```
Comment (shwetathareja) [inline libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java:41]: With all the methods exposed in SpanScope, we expect to have scope object explicitly and don't expect magic through threadContext? Do we still need to set it in threadContext?

**Review by shwetathareja — COMMENTED** [review]
Thanks @suranjay for the refactoring the ```SpanScope``` class.

File: libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java (hunk 3 above)
Comment (suranjay) [reply libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java:41]: Yes. We need thread context to save the current active span to be used as parent for any new span

**Comment by suranjay** [comment]
> > @suranjay @Gaganjuneja seems like folks you are have conflicting changes #8357
> 
> Yes, we will rebase the other PR changes which gets merged earlier

Merged changes from @Gaganjuneja commit. @reta Please have a look when you get chance

**Review by reta — APPROVED** [review]

File: plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelSpan.java (hunk 1 above)
Comment (suranjay) [reply plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/OTelSpan.java]: Not needed now as we have changed the code

**Review by shwetathareja — APPROVED** [review]

**Pull request merged**